Tools for Measuring Molecular (Tumor) Evolution
===============================================

_Suraj Joshji_ (April 28, 2022)

Tools discussed:

**Clustering**

- [GitHub - Roth-Lab/pyclone: Probabilistic model for inferring clonal population structure from deep NGS sequencing.](https://github.com/Roth-Lab/pyclone)
- [GitHub - genome/sciclone: An R package for inferring the subclonal architecture of tumors](https://github.com/genome/sciclone)

**Tree-building**

- [GitHub - morrislab/phylowgs: Application for inferring subclonal composition and evolution from whole-genome sequencing data.](https://github.com/morrislab/phylowgs)
- [GitHub - raphael-group/calder: CALDER (Cancer Analysis of Longitudinal Data through Evolutionary Reconstruction) reconstructs evolutionary trees from longitudinal bulk DNA sequencing data](https://github.com/raphael-group/calder)
- [GitHub - amcpherson/citup: Clonality inference in multiple tumor samples using phylogeny](https://github.com/amcpherson/citup)

**Visualization**

- [GitHub - YuLab-SMU/ggtree: Visualization and annotation of phylogenetic trees](https://github.com/YuLab-SMU/ggtree)
- [GitHub - chrisamiller/fishplot: Create timecourse "fish plots" that show changes in the clonal architecture of tumors](https://github.com/chrisamiller/fishplot)
- [ETE Toolkit - Analysis and Visualization of (phylogenetic) trees](http://etetoolkit.org)

